### ISSUE TITLE
Move parser choice to config

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#1749]
I think we might of missed an opportunity in `7.0` to handle parsers in a more extensible way. I think we should fix this in `8.0`. 

Sharable configs, plugins and (in `7.0`) processors are all part of our design philosophy to keep development on stylelint sustainable by putting systems in place that allow the community to extend a robust core, rather than have us bolt more things onto core.

I think we should treat parsers in the same way e.g.

`npm install postcss-stylus --save`

``` json
{
  "parser": "postcss-stylus",
  "processors": "stylelint-processor-markdown",
  "extends": "stylelint-config-standard",
  "plugins": [
    "../special-rule.js"
  ],
  "rules": {
    "plugin/special-rule": "everything"
  }
}
```

It's [this discussion in stylint (a linter for stylus)](https://github.com/rossPatton/stylint/issues/303#issuecomment-236128283) that made me realised we might of made a mistake by pulling `postcss-scss`, `postcss-less` and `sugarss` into core.

In `7.0` we automatically set the parser based on the file extension. This was primarily to reduce the number of new issues we were getting. I think relying on a [helpful warning](https://github.com/postcss/postcss/issues/856) would suffice. e.g. "You are attempting to lint a .scss file with the standard parser. Install `postcss-scss` and change your configuration object to use it.". This feels like a fair compromise for extensibility.

Moving parsers outside of core will also help more clearly define what is the responsibility of the stylelint org and what belongs to the parser authors.

I'm in favour of nipping this one in the bud as it feels like an important step in ensuring development on stylelint remains sustainable. We've not been afraid to make these kind of breaking changes in the past when we're confident of the benefits they'll bring later down the line. I think stylelint is a better tool because of these choices. Thoughts?

**Comment by davidtheclark** [comment]
I worry that if we open the door to *any* parser, we open the door to a confusing, possibly unanswerable set of questions and "bugs" related to what stylelint "supports". I worry that that would actually cause *more* headaches than the setup we currently have, with a finite, controlled list of support syntaxes.

PostCSS provides an AST that stylelint starts with; but there are no guarantees about what that AST contains, so there's no guarantee that any arbitrary parser will actually meet stylelint's expectations. Moreover, a *huge* amount of string analysis is performed outside of PostCSS — within stylelint code and in supporting tools like the selector and value parsers. Just because a parser is written that makes PostCSS produce an AST from a string does not mean (A) that stylelint will understand the AST that it produces, or (B) that stylelint rules will be able to deal with that string or its substrings.

Right now we are actively trying to *support* SCSS, Less, and SugarSS syntaxes. We expend energy ensuring that stylelint won't stumble on those syntaxes, and try to make as many rules as possible work for them. We are *not* trying to support Stylus; so nobody is going to open issues complaining about problems using stylelint with Stylus. My guess is that if someone does write a `postcss-stylus` parser and tries to use it with stylelint, she'll encounter *lots* and *lots* of problems, with both the AST and string analysis. I'd expect that stylelint "supporting" Stylus would not be nearly so simple as plugging in a `postcss-stylus` parser.

I worry that suggesting anybody can plug in an arbitrary parser will not offload work from stylelint core into userland, but will in fact increase the burden on stylelint core. If a user plugs in `postcss-stylus` and encounters all kinds of errors, what is she going to do? She's going to open issues in stylelint. What else *could* she do?

**Comment by ntwb** [comment]
Reading @jeddy3's issue I thought that sounds reasonable, then I read the @davidtheclark's reply and I think there are some important concerns there, if stylelint has a free-for-all approach in allowing any parser then as pointed out a greater burden will be placed on stylelint core whether it's warranted or not.

**Comment by jeddy3** [comment]
@davidtheclark All good points. And @ntwb, I agree they are important concerns. However, I'm going to try to argue that my proposal will mitigate this concerns, rather than exacerbate them.

> I worry that if we open the door to any parser, we open the door to a confusing, possibly unanswerable set of questions and "bugs" related to what stylelint "supports".

The door is already open. When stylelint is [used](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/postcss-plugin.md#example-a) as a PostCSS plugin, whether via the PostCSS JS API or via its many tasker runner plugins, an arbitrary parser can be used. So, at the moment we're trying to *actively* support SCSS, Less, and SugarSS syntaxes *and* the door is open to arbitrary parsers. So, if someone does write `postcss-stylus` and they try to use it with the PostCSS plugin, then we're already in position where they're going to open issues in this repo.

I don't think this is really about mitigating some future problem though, but a problem we have now. I think what we lack is a means to consistently and clearly define what the stylelint org "supports". I think, for example, having the SCSS parser in core, but having `stylelint-scss` in the community creates ambiguity.

I think by having only the default PostCSS parser in core, and moving non-standard parsers to the community, we can remove that ambiguity.  It also forces a conscious decision by the user to *opt-out* out of core support and into that of the community.

Inside the [configuration guide](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md), just before [plugins](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#plugins), we can document our support policy in a way that's consistent with other parts of the configuration guide e.g. (borrowing heavily from [eslint](http://eslint.org/docs/user-guide/configuring#specifying-parser)):

---

### Parser

By default, stylelint uses the standard PostCSS parser. You can optionally specify that a different, community-supported, parser should be used in your configuration object so long as the parser meets the following requirements:

1. It must be a PostCSS-compatible parser.
2. It must produce a stylelint-compatible AST.
3. It must produce strings within the AST that are stylelint-compatible i.e. compatible with `postcss-selector-parser`, `postcss-value-parser`, `postcss-media-parser` and `style-search`.

Note that even with these compatibilities, there are no guarantees that an external parser will work correctly with stylelint and the stylelint org will not fix bugs related to incompatibilities with other parsers. The onus of addressing any incompatibilities in the generation of the AST is on the parser authors. Having said that, we welcome to contributions to the `isStandardSyntax*` utils. These utils help rules avoid *tripping up* on non-standard syntax and constructs.

To specify a community-supported parser, specify the name of the npm module as the value of the `parser` property within your configuration object. For example, to use the `postcss-scss` parser:

```json
{
  "parser": "postcss-scss"
}
```

The following parsers are known to be, in varying degrees, compatible with stylelint:

- `postcss-scss`
- `postcss-less`
- `sugarss`

However, non-standard constructs introduced by these language extensions are ignored by stylelint's rules.

If you want we use all the rules of stylelint and be confident they'll work with every line of code, we recommend writing CSS code that uses only standard constructs and using the standard parser.

---

I believe this makes it clearer where the responsibilities lie. I think it also minimises the risk of moving the parser choice to the config being a "free-for-all" approach. In the same way that `processors` aren't arbitrary (but rather they must be compatible with stylelint), the criteria for useable parsers is documented in a clear and concise way.

Even if `postcss-stylus` is never written, this approach feels like it clarifies our position on SCSS, Less and SugarSS. Take https://github.com/stylelint/stylelint/issues/1741 for example, if we had this system in place we could close that and say it's the parser author's responsibility. Right now, it feels like we need to keep that bug open *because* we (the stylelint org) are *actively* trying to support Less. 

I think this proposal frees up the stylelint org from *actively supporting* these parsers by making it easier to communicate that only standard-syntax is the responsibility of the stylelint org. Non-standard syntax is the responsibility of the wider community and is done with community plugins like `stylelint-scss` and community parsers like `postcss-scss`. That's not to say work won't continue on the `isStandardSyntax*` utils, but the remit of work within the org is clearer i.e. the `isStandardSyntax` utils will be a well-tested way of ignoring non-standard syntax and constructs when attempting to use the core rules.

These points flesh out the original idea that such an approach would create a more robust and sustainable core. If I'm using vanilla CSS in my project (i.e. not using SCSS or Less) then I have access to a lean tool. If, for example, I'm using SCSS then I can *opt-in* and bolt on the necessary bits e.g:

```json
{
  "parser": "postcss-scss",
  "plugins": ["stylelint-scss"]
}
```

I think vanilla CSS is going to be around for long time, but monolithic preprocessors will come and go. I think we've an opportunity here to focus stylelint on the former while putting in place an extensible system (that be clearly communicated) for the latter.

**Comment by evilebottnawi** [comment]
@jeddy3 what about if you have two syntax, right now we migrate from scss to cssnext, but have many code on scss and unfortunately hard to make the transition quickly :disappointed: . Be good, if we was allowed setting array of parsers.

**Comment by davidtheclark** [comment]
@evilebottnawi You could always run two stylelint tasks, one for SCSS and one for CSS.

@jeddy3 I'm afraid I still don't see the concrete benefit of this; it still looks like added confusion to me.

This is what I understand our position to be right now: stylelint core rules target standard CSS constructs. stylelint can *run* on stylesheets using the non-standard syntaxes SCSS, Less, SugarSS, but typically core rules simply any ignore non-standard constructs. Any rules specifically targeting non-standard constructs exist as plugins, outside of stylelint core. 

To put it another way: Within stylelint core, our goal related to a limited number of non-standard syntaxes is for stylelint to not fail on those stylesheets. I think that's a clear and limited goal *that expands our potential user-base tremendously*.

I'm trying to think through a scenario played out under the new regime you propose, and I don't see it working. Let's say Tom uses SCSS maps. He tries to use rule X but hits an error that is caused by the fact that he uses SCSS maps. What does he do? Currently, we'd suggest that Tom make a PR to stylelint ensuring that rule X does not outright fail when it encounters SCSS maps, and if Tom wants to enforce something related to those maps he should make a PR to `stylelint-scss`. Under the new regime, what would he do? He'd simply be unable to use rule X? Or `stylelint-scss` would have to *duplicate* rule X with some added logic to avoid tripping on SCSS maps?

I also worry that the suggested rules determining "stylelint-compatible" parsers are unrealistic. "It must produce a stylelint-compatible AST." We don't even really know what that means. "It must produce strings within the AST that are stylelint-compatible." That depends entirely on the source language, not the parser.

**Given the present scenario, described above, what are the problems that we want to solve?** Let's try to articulate those problems and then ask if this proposal solve them, or if there is something else we could do?

## Problem: We don't want people to think stylelint core is responsible for all the possible non-standard syntaxes under the sun.

I think the clearest way to do this is to maintain our current principle ("our only goal related to non-standard syntaxes is for stylelint to not fail on those stylesheets") with a limited number of non-standard syntaxes: SCSS, Less, and SugarSS. I think our position here is already clear.

> When stylelint is used as a PostCSS plugin, whether via the PostCSS JS API or via its many tasker runner plugins, an arbitrary parser can be used.

True, and I guess we should just make it clear that there are no guarantees if you don't use one of the supported syntaxes. Easy as that. 

A statement like "The following parsers are known to be, in varying degrees, compatible with stylelint" really underplays the amount of effort we've put into *and success* we've had with the  supported syntaxes. Putting `postcss-scss` on the same level as a hypothetical `postcss-stylus` seems like a big mistake to me.

## Problem: We don't want to expend a lot of energy fixing errors caused by non-standard syntaxes.

How about just not expending a lot of energy fixing those errors? We've already made a *ton* of progress, and I have the impression now that our support for the non-standard syntaxes is actually pretty good. *We do not need to move at the rate we are moving.* And *we do not need to fix every bug within a week.*

Another possibility is to state outright that when there's a problem with non-standard syntax, we only accept issues accompanied by PRs. Essentially, we'd be saying that the people using non-standard syntaxes should do the work of ensuring stylelint works with it. (It would be great if we had a champion of each syntax who liked tackling issues related to it.)

Ok, I've typed enough here. Here's what I'm saying: In order to be convinced of the parser-in-config proposal, I need to see more clearly what problems we have that this will solve, how it will solve them, and how its solutions are more satisfactory than other possible solutions.

**Comment by sendilkumarn** [comment]
valid points 👍 

IMO, we should start with what is stylelint's compatible AST. And with that we can have a layer in between which handles these non-standard
So the core remains intact and this layer takes non-standard syntax and does any parsing / validating that is required before checking the rule with the core.
![untitled diagram](https://cloud.githubusercontent.com/assets/12471122/17356508/672aae76-598a-11e6-9f54-64a0f1347102.png)

It will be more modular. Suppose in future if there some non-standard syntax is becoming obsolete, then we need not carry that along the framework. we can simply remove this layer in between. 

It will also always prevents stylelint's main intent of linting css.

**Comment by jeddy3** [comment]
> It will be more modular. Suppose in future if there some non-standard syntax is becoming obsolete, then we need not carry that along the framework. we can simply remove this layer in between.

This is exactly the crux of what I'm trying to achieve with this proposal. I think I muddied the waters somewhat with unclear writing (especially within that documentation suggestion). @davidtheclark I'm not suggesting a regime change. Like you, I think what we have now is a clear and limited goal, and I agree that it expands our potential user-base tremendously. I also think how issues are currently dealt with works well. What I'm proposing is a tweak in how the non-standard parsers are used; from being bundled within stylelint to being bolted-on (when needed) by the user. 

This isn't really to fix a problem we have now, but one I'm anticipating we might have a year or two down the line. The modular philosophy is something that's worked well for us so far. Adopting a parser-in-config approach now, while the user-base is still fairly young, feels like a continuation of that philosophy.

> A statement like "The following parsers are known to be, in varying degrees, compatible with stylelint" really underplays the amount of effort we've put into and success we've had with the supported syntaxes. 

Agreed. It wasn't my intent to underplay it. I originally thought of some kind of support-gauge that shows that SCSS non-standard constructs rarely cause issues now, Less ones only a little more often etc. It was lazy writing.

**Comment by davidtheclark** [comment]
@sendilkumarn The proposal is a fine idea, but a huge amount of work, a very significant layer of additional abstraction, for problems we're not sure we have — right? I guess I'd be a lot more interested in the idea, too, if I could be convinced by some code implementing it.

> This isn't really to fix a problem we have now, but one I'm anticipating we might have a year or two down the line.

What exactly are the problems we're anticipating? That another preprocessor with all kinds of wacky new non-standard syntax arises and gains popularity? That Sass popularity dwindles and we feel like we can drop support? I still think we need to more clearly articulate the problems so we can assess what level of energy is worth investing. Maybe I'm misunderstanding, but I have the impression that what you're both suggesting here is an *enormous* task, so it needs to be very well justified.

**Comment by sendilkumarn** [comment]
Something like this. 

`comments-newline-after`
`value-list-comma-newline-after`  -->  consider this two for example. _[ just for example ]_

we can merge this into *newline-after rule which will be in `stylelint core`

But comments / value - list  should be checked / validated in the middle layer `css/scss/non-standard syntax`

I will try to create a PR for this. If my understanding is correct

**Comment by jeddy3** [comment]
> I will try to create a PR for this. If my understanding is correct

Yes please, as I'm now a little confused by what you're proposing.

**Comment by ntwb** [comment]
I've not got much to add to the core issue here, and no strong opinion either way, much of the work that would be involved I'd say is technically over my head.

That said a couple of concerns in regard to some of the non-technical issues raised:

Currently stylelint is attracting SCSS, LESS, SugarSS, and Stylus users, sadly Stylus users are turned away as there is no `postcss-stylus` parser at this stage. In stylelint continuing the be the "central hub" of all things stylelint we're able to steer people in the right direction, Davids use case outlined above with Tom and hist SCSS is a good example of this I think, we'd be driving Tom away from the get go rather than helping steer him to `postcss-scss`, or `stylelint-scss` as necessitated, hopefully gaining Tom as a valued contributor to stylelint and what I'd call stylelint's "sister projects" `postcss-scss` and `stylelint-scss` as a contributor. 

stylelint has an excellent _ecosystem_ and which that in part is what I believe Richard is proposing here to improve further. But by doing so does that fragment the stylelint community by having isolated stylelint communities from one another, a stylelint CSS community, a stylelint SCSS community, a stylelint LESS community and so on rather than one all encompassing stylelint community?

> _We do not need to move at the rate we are moving. And we do not need to fix every bug within a week._

The pace stylelint moves at is very fast, I don't think I need to link to the numerous times this has been mentioned over recent months but time and time again to paraphrase "WIth all this awesome work in x.y release let's take a few days off" and not long after the pace is once again _flat chat_.

I think that by deliberately slowing down development pace would achieve a few things, it would allow the ~60 watchers and ~2000 people who have starred the repo the opportunity to contribute more, rather than a stylelint member commenting and creating pull requests instantly which is not a bad thing, it can however not leave any opportunity for others to get involved, typically I'd say this occurs with the "easy" to fix bugs, the harder items can and do sit around a little longer. 

By allowing issues to have more "time", would allow the stylelint team some time to relax, pull requests can "soak" a little longer before being merged allowing further time for feedback, even if to some extend they are only "plus ones".

**Comment by jeddy3** [comment]
@ntwb Thanks for sharing your thoughts!

> I don't think I need to link to the numerous times this has been mentioned over recent months but time and time again to paraphrase "WIth all this awesome work in x.y release let's take a few days off" and not long after the pace is once again flat chat.

Yep, that was my bad. I knew I was heading back to London for work and so I tried to get as much done before I did. I think the last couple of weeks have been at a far more sustainable pace, and I think the pace will continue to slow.

The positive aspects you highlight of slowing the pace are great. A number of first-time contributors tackled issues for `7.2.0`, and I suspect that's partly because of the slower pace.

---

Back to the issue in hand, after re-reading @davidtheclark's arguments I'm now unsure the disruption caused by moving the parser choice to the config is worth it, nor the effort required. At least not in the way [I suggested](https://github.com/stylelint/stylelint/issues/1764).

As @ntwb suggested, there's no harm in leaving this issue stew for a while though. Perhaps @sendilkumarn will propose a PR that helps us understand the *layered* architecture a little better.
